design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,381 @@
+    async fn fetch_chunk(
+        &self,
+        chunk_location: &str,
+        range: &ByteRange,
+        checksum: Option<&Checksum>,
+    ) -> Result<Bytes, VirtualReferenceError>;
+}
+
+```
+
+We will have one implementation for each `ObjectStorePlatform` where we want to "speak" a different protocol or optimize differently. For example, we can have one implementation that uses the Rust native S3 client for `s3` and other one that uses object_store for `LocalFileSystem`.
+
+The `Repository` now owns a `VirtualChunkResolver`. This struct is in charge of hosting the containers, matching them and building the `ChunkFetchers` needed.
+
+```rust
+pub struct VirtualChunkResolver {
+    containers: Vec<VirtualChunkContainer>,
+    credentials: HashMap<ContainerName, ObjectStoreCredentials>,
+    fetchers: RwLock<HashMap<ContainerName, Arc<dyn ChunkFetcher>>>,
+}
+```
+
+Creating a new object store client for each virtual chunk request would obviously not work well. We'll have a cache of clients, one per `VirtualChunkContainer` in use. This is not the most efficient thing, we could share the same client across containers that share the platform for example, but it's good enough and simpler. We can always optimize later.
+
+The cached `ChunkFetchers` are maintained by the `VirtualChunkResolver` in its `fetchers` field. Fetchers are instantiated on demand, on the first chunk that needs them and then cached without invalidation. Careful coding is required to avoid a herd effect when multiple chunks for the same container are requested concurrently (a very common situation)
+
+### Credentials
+
+We haven't discussed the third field in the `VirtualChunkResolver`: `credentials`. This field gets populated when the `Repository` is initialized and it holds the credentials passed by the user for each container. Users don't necessarily need to pass credentials for every container, but if they don't, their requests for those chunks may fail.
+
+The `ContainerName` field in the `VirtualChunkContainer` is how the users refer to them to pass credentials.
+
+Several types of credentials are possible.
+
+```rust
+pub enum ObjectStoreCredentials {
+    FromEnv,
+    Anonymous,
+    Static {
+        access_key_id: String,
+        secret_access_key: String,
+        session_token: Option<String>,
+    },
+}
+```
+
+An we will have to add at least one more variant once we implement credentials refresh.
+
+### Use case examples
+
+#### Basic usage
+
+All virtual chunks point to S3 us-east-1, single credential set.
+
+* When calling `set_virtual_ref` locations are passed with `s3://` protocol.
+* No persistent configuration needs to be set in the repo.
+* At runtime, the repo instance can be configured passing credentials for container "s3", the default container.
+* If the repo data is also in s3 us-east-1, and the same credentials can be used, the previous point is not needed, the default will be using the same credentials as the ones for the repo.
```

Comment:
I don't understand why reusing the credentials requires it to be in the same region. This feels arbitrary.

Credentials are not scoped to regions or even buckets.